### ISSUE TITLE
API is a bit strange

### ISSUE THREAD

**Issue by dAnjou** [sloria/environs#128]
I know the title is a bit weird as well but bear with me, I'll explain where I'm coming from.

Instances of classes usually hold and manage state related to their name. The main class in this library is called `Env`. That suggests to me that an instance of `Env` holds and manages the state of the environment. Now, let's look at some code:
```python
>>> from environs import Env
>>> import os
>>> env = Env()
>>> env('FOO')
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "/home/user/project/env/lib64/python3.7/site-packages/environs/__init__.py", line 74, in method
    raise EnvValidationError('Environment variable "{}" not set'.format(source_key), [message])
environs.EnvValidationError: Environment variable "FOO" not set
>>> os.environ['FOO'] = 'bar'
>>> env('FOO')
'bar'
```
To me that is quite unexpected. It looks like manipulating `os.environ` changes the state of `env`, that's not really nice behavior.

I do know that this is actually not how the library works. `Env` does not actually hold and manage any state related to the environment. The only state it holds is the _parsing state_, as in the variables it has parsed so far. Let's look at some more code (continuing where I left off):
```python
>>> env.dump()
{'FOO': 'bar'}
>>> os.environ['FOO'] = 'lorem'
>>> env.dump()
{'FOO': 'bar'}
>>> env('FOO')
'lorem'
>>> env.dump()
{'FOO': 'lorem'}
```
`dump` seems to return all variables that have been parsed so far, including their values. But after changing the environment it still has the old value, quite unexpected again, it does keep some state. Only after parsing again, the new value shows up in the dump.

Another confusing behavior is that of `read_env` which does not suggest any side effects but it does change the current environment. Granted, it does not override anything but it's still not ideal behavior.

Some of the confusion can probably be fixed with better documentation.

However, the reason I started writing this is another one. I was hoping that this library could help me parse the environment of another system, specifically a Docker container. But while it has all the wonderful tools I can't do that without polluting my own environment, and `read_env` only accepts file system paths while the underlying library would happily accept file-like objects as well.

**Comment by sloria** [comment]
Right, `Env` dynamically builds an internal schema and keeps the state of the parsed variables. Perhaps a bit strange, but I think it's not so bad--the ergonomics are nice in the end. 

> Another confusing behavior is that of read_env which does not suggest any side effects but it does change the current environment. Granted, it does not override anything but it's still not ideal behavior.

This is the behavior of the underlying library, python-dotenv, so that's outside the scope of environs.

> Some of the confusion can probably be fixed with better documentation.

PRs are welcome =)

**Comment by epicserve** [comment]
Seems like this could be closed, unless others are in agreement that this type of change is needed.

**Comment by sloria** [comment]
OK, closing for now. Feel free to continue discussion here, though.

**Comment by dAnjou** [comment]
> Right, `Env` dynamically builds an internal schema and keeps the state of the parsed variables. Perhaps a bit strange, but I think it's not so bad--the ergonomics are nice in the end.

Maybe I'm alone with this but, like I demonstrated, it was quite unintuitive to me, so I wouldn't call the ergonomics nice.

> > Another confusing behavior is that of read_env which does not suggest any side effects but it does change the current environment. Granted, it does not override anything but it's still not ideal behavior.
> 
> This is the behavior of the underlying library, python-dotenv, so that's outside the scope of environs.

Not quite true, or at least not addressing the issue (which I might have misphrased). The moment you're wrapping another libraries functionality and make it public it becomes your functionality and, in fact, _is_ in scope. Also, the name in python-dotenv is `load_env` which has quite a different meaning than `read_env`.

> > Some of the confusion can probably be fixed with better documentation.
> 
> PRs are welcome =)

I don't feel invested enough to spend more time on this because I was only evaluating this library and eventually decided against it because of its API and because I needed something slightly different (i.e. parsing the output of the `env` command which python-dotenv's `load_env` allows via its `stream` argument but environ's `read_env` does not accept). And I'd much rather see an improved API in a new major version.

However, I do think this library has some nice features which is why I wrote down my thoughts in this issue. I appreciate that you took the time to read it.
